src/main/java/com/google/api/generator/engine/ast/TypeNode.java

```diff
@@ -88,6 +88,31 @@
   @Nullable
   public abstract Reference reference();
 
+  @Override
+  public int compareTo(TypeNode other) {
+    // Ascending order of name.
+    if (isPrimitiveType()) {
+      if (other.isPrimitiveType()) {
+        return typeKind().name().compareTo(other.typeKind().name());
+      }
+      // b is a reference type or null, so a < b.
+      return -1;
+    }
+
+    if (this.equals(TypeNode.NULL)) {
+      // Can't self-compare, so we don't need to check whether the other one is TypeNode.NULL.
+      return other.isPrimitiveType() ? 1 : -1;
+    }
+
```

Comment:
Do we need to consider the VOID type comparsion?

Reply:
No, it doesn't need to be handled separately.